### ISSUE TITLE
Filtering plugin and Tile.drawCanvas cache warning

### ISSUE THREAD

**Issue by pearcetm** [openseadragon/openseadragon#2277]
This is related to https://github.com/openseadragon/openseadragon/issues/1421 and https://github.com/openseadragon/openseadragon/issues/2054.

I am seeing a console warning sporadically along the lines of `[Tile.drawCanvas] attempting to draw tile 8/130_143 when it's not cached`. This is happening while using the [OpenSeadragon Filtering plugin](https://github.com/usnistgov/OpenSeadragonFiltering).

Can this be safely ignored? Can the warning be disabled? As discussed in the linked issues, upping the cache size might help, but I'm not sure that's necessarily the right way to go... Are tiles that aren't cached eventually processed/drawn? The `drawCanvas` function returns immediately after throwing this warning, so I'm not clear exactly how the draw logic is implemented once this happens.

**Comment by Aiosa** [comment]
It sure looks like an issue with the filtering plugin. The cache API in OpenSeadragon has changed a bit since #1421, and seeing this issue is still relevant makes me really think the plugin does something in the wrong way. I tried to skim through it quickly but did not identify any suspicious areas. 

It is not safe to ignore the warning, the tile draw event exists early because no data is available. The cache system discarded old data and did not load it back when the viewport returned to a position where it is needed again. Increasing the tile cache is a possible hotfix, but not really something you should rely on. It just means it is rarer for tiles to get discarded. The cache should remove both data and the tile instance so a tile not having cached data means there is some error, possibly caused by the plugin. The issue should be taken care of by the plugin developer(s). 

Another hotfix could be to bypass the cache fully, by saving `tile.context2D = context;` instead https://github.com/usnistgov/OpenSeadragonFiltering/blob/f1e350c08e8bf050fdc84b89e9b04ad041c861b9/openseadragon-filtering.js#L91 - this would turn off caching, but you would have to adjust the code to correctly in `tileDrawingHandler` to update the data of `rendered` properly, since `rendered = tile.context2D` in this case; so you would ultimately want to create _two_ canvas contexts (one with raw data one with filter result)...

I am thinking about releasing my [webgl OSD postprocessing](https://github.com/RationAI/xopat/tree/master/modules/webgl) as a plugin too, just did not have time to tidy it up. If you were interested and needed some help with it you can let me know.

**Comment by iangilman** [comment]
I'm not sure exactly what's happening here, but it does seem like a sign of something going wrong. Like @Aiosa says, I think if the tile got purged, the whole tile would be gone, not just its cache. Do you get the same thing happening without the plugin? If not, that would definitely point to the plugin. To track it down, you might look for places where the plugin does things with `context2D` or `cacheImageRecord`. 

On the other hand, does it cause any visual issue in the viewer? If not, then I guess you can ignore it... I wouldn't necessarily want to remove it from OSD, though.

**Comment by pearcetm** [comment]
I confirmed that it only happens while using the filtering plugin. Setting the maxTileCacheCount to something low (e.g. 50) makes it very easy to test. When the cache is this small, it also makes for some very weird visual effects! The filter kind of flickers on and off a bunch of times on a tile-by-tile basis and tons of warnings are raised. Ultimately the whole viewer does render... it kind of seems like there's a race condition somehow between the async filtering operation and the cache/drawer behavior. I haven't fully tracked down the root of it, but I'll keep digging.

@Aiosa I'd be interested in giving your webgl implementation a try. I guess the best way to get started would be to fork your repo... I'll likely ping you for advice on how to use it properly once I get to that point.

**Comment by Aiosa** [comment]
@pearcetm yes, it is possible that the process of per pixel JavaScript processing just can't keep up with such a small cache, being forced to render too often.

Forking the repo might not be the best idea since it contains a whole JavaScript app, you just need the directory in the link I gave. As I said, did not have time to split it yet. Mainly because I have bad experience with git submodules and don't know how to approach it effectively.

**Comment by pearcetm** [comment]
I'm still not sure what's going on, but I've tracked the issue down a bit further...

The problem is NOT with javascript processing of canvas data (in my current case, at least). The issue arises when a tile is cached twice, by the plugin and by the default action, which then causes the warning/problem.

In `tiledimage.js`:
```
        this.viewer.raiseEvent("tile-loaded", {
            tile: tile,
            tiledImage: this,
            tileRequest: tileRequest,
            get image() {
                $.console.error("[tile-loaded] event 'image' has been deprecated. Use 'data' property instead.");
                return data;
            },
            data: data,
            getCompletionCallback: getCompletionCallback  //<---- callback is obtained and called by handler (first time)
        });
        // In case the completion callback is never called, we at least force it once.
        getCompletionCallback()(); // <----- callback obtained and called by default (second time)
```

Circumventing this duplicate caching in various ways prevents the issue and the viewer works great in this case. But those hacks aren't exactly robust. Given this behavior I've observed, I suspect there's something funky with the cache when the same tile is cached twice, but I'm still trying to figure out what's actually going on. Following the logic of all of this is not straightforward unfortunately :(

**Comment by Aiosa** [comment]
Should be irrelevant. `tile-loaded` event just enable the plugin to attach custom data to the tile instance, while `tile-drawing` event is called on the tile instance when it is about to be drawn. Should a tile be discarded (cache erased), `tile-drawing` should be never fired on the tile instance from that point on - whether it was or was not freed from the memory -> OSD should no longer know how to access it, and so never fire the event again. It would instead create new tile, call again loaded and then drawing. So even if the plugin kept a reference to the old tile with erased cache, it means nothing.

**Comment by iangilman** [comment]
Is the plugin calling `getCompletionCallback`? If so, is it calling it in the event handler or is it doing so asynchronously? If it's doing it asynchronously, that would account for it getting called twice; it needs to be called directly in the event handler. Perhaps this needs to be documented better.

**Comment by pearcetm** [comment]
@iangilman `getCompletionCallback` is called synchronously (incrementing the counter), but then the resulting callback is also executed synchronously (decrementing the counter back to zero, and caching the tile). Execution proceeds to the next line, which does the same thing - get and call the callback. So the counter goes 0 -> 1 -> 0 *cache tile* -> 1 -> 0 *cache tile*

In `cacheTile`, two `TileRecord`s are created, both referencing the same underlying `tile` (and taking double the space in the cache). When the cache fills up, one of these is unloaded (in `_unloadTile`) but the other stays in; `tile.cacheImageRecord = null` is called, and this applies to the one still in the cache (again, because the same tile is referenced twice). I suspect but haven't had a chance to prove that this is the root of the problem.

**Comment by pearcetm** [comment]
Ok, if the following two things are true, the cache works properly:
1) `getCompletionCallback()` *must* be called **synchronously**
2) the `callback` function that is returned *must* be called **asynchronously**

This is understandable if you know how the increment/decrement logic is implemented. At a minimum I think it needs to be better documented. Something along the lines of

>  If tile data is modified synchronously, the callback must not be used. If asynchronous processing is desired, `getCompletionCallback()` must be  called **synchronously**, and the callback executed **asynchronously**. If `getCompletionCallback()` is called, the callback **must** be executed.

 In the context of the plugin, that means even if a custom filter is to be executed synchronously, the callback must be deferred by `window.setTimeout` or similar.

Even though this does work as intended if used just right, it seems kind of fragile (and the warnings that are generated are not particularly helpful in tracking down the root cause). @iangilman @Aiosa what would you think about trying to make `cacheTile` and related functions more robust?

**Comment by Aiosa** [comment]
Oh I did not notice the plugin using `getCompletionCallback` - honestly, I think removing its usage would fix everything - the cache is not used at the time so why calling the completion callback. It's still true that calling the callback immediately makes the counter go 0->1->0 so each such execution will call the callback again and again. It at the end looks like a bug in OSD. Wow.

**Comment by pearcetm** [comment]
> Oh I did not notice the plugin using getCompletionCallback - honestly, I think removing its usage would fix everything - it is not async so why calling the completion callback.

The plugin is supposed to be flexible, in that some filters (computationally cheap and fast) should be allowed to be synchronous, while other slower ones should be allowed to be asynchronous, and multiple filters applied at the same time. This flexibility also lets it work with other image processing libraries that provide sync/async functions. The way it's implemented currently (which goes back 5+ years and is supposed to be compatible with OSD v2.1+ per the plugin readme), it always calls `getCompletionCallback` and expects the filter(s) to each call the callback to indicate that processing is complete.

My current understanding is that the purpose of `getCompletionCallback` is primarily so that caching (and rendering) can be delayed until all processing is completed for a given tile, but I am certainly not an expert - I just started looking into this in the past couple of days.

Would it be a problem to just check whether a tile has been cached already, before adding it to the cache? `tile.cacheImageRecord` is set when the tile is cached, and unset during `_unloadTile`. We could take advantage of this within `cacheTile` and just return immediately if `tile.cacheImageRecord` is truthy.

**Comment by Aiosa** [comment]
> 
> My current understanding is that the purpose of `getCompletionCallback` is primarily so that caching (and rendering) can be delayed until all processing is completed for a given tile, but I am certainly not an expert - I just started looking into this in the past couple of days.
> 

The whole event comes as a callback from tile fetching so it is (should be) asynchronous wrt. OSD as such. Since the plugin does everything without async, it does not need to call the completion at all, it will be called within the fallback after event invocation. Allowing async here is more of a compatibility issue. If you consider using a third party API in this event that runs async, you need the callback (or synchronized async event handlers) to say 'ready now'. Otherwise you execute synchronously and 'now' is implicit. It should not matter whether the handler is async in this case, as the tile is ready to render once the completion is called. Implicitly or explicitly.

**Comment by Aiosa** [comment]
Huh only now I've noticed the completion callback getter docs really mentions it should be async... I guess more robust implementation wouldn't hurt.

**Comment by pearcetm** [comment]
To the extent that there's a bug, it's that the plugin always calls `getCompletionCallback` even if there are no asynchronous filters. The documentation for the `tile-loaded` event does indicate that this can be called to indicate that asynchronous processing is complete:

> A function giving a callback to call when the asynchronous processing of the image is done.

However, it is not entirely clear from the wording that the processing **must** be asynchronous. The error made by the plugin falls through silently for a while, and eventually causes the seemingly unrelated warning that this whole issue got started with. So I think that if we want to continue to use the current code, updating the documentation and error/warning messages would be good, and I can do so once we settle on an approach.

That said, I think it would be better to allow the callback to be called either `sync` or `async`. Given the current logic of caching and unloading tiles, I do not think it is ever intended for a tile to be cached twice (once it's been unloaded it can be cached again of course; I mean twice in the same active cache). I can imagine a couple of approaches that would work that would avoid duplicate caching. Logic could be added to the `getCompletionCallback` function to allow the callback to only be called once (instead of every time the counter drops to zero), so that the tile is never passed more than once into `cacheTile`. `cacheTile` could check whether the tile has already been cached (https://github.com/openseadragon/openseadragon/pull/2283), which means that subsequent requests to cache the tile are simply ignored. I believe these would essentially be equivalent, and both would allow the callback to be called any time.

**Comment by iangilman** [comment]
`getCompletionCallback` is an unusual API, but I think it does a pretty good job of accomplishing its mission, aside from these bugs we are finding. I propose we add two more flags: One to keep track of whether the completion has happened and another to keep track of whether we're past the point of being able to reasonably call `getCompletionCallback`. Something like so: 

```
        var increment = 0,
            completed = false,
            tooLate = false,
            _this = this;

        function getCompletionCallback() {
            if (tooLate) {
                $.console.error('You must call this synchronously!');
            }
            increment++;
            return completionCallback;
        }

        function completionCallback() {
            increment--;
            if (increment === 0 && !completed) {
                completed = true;
                ...
```

... And: 

```
        ...
        // In case the completion callback is never called, we at least force it once.
        getCompletionCallback()();

        tooLate = true;
```

That helps us enforce that `getCompletionCallback` is called synchronously and makes sure the actual completion doesn't happen twice. It also has the advantage of fixing the issue with the filtering plugin without requiring the plugin to change.

That's just a quick draft... Maybe the variable names could be better and certainly the error message should be more intelligent. 

What do y'all think? Thank you for the in-depth analysis here!

**Comment by Aiosa** [comment]
Hmm, I indeed sometimes try to complicate stuff too much. Changed my PR to the flag-guard version. https://github.com/openseadragon/openseadragon/pull/2282. It allows early completion too, although there might be problem that some plugin will use `context2D` param but the completion callback gets called before that so the cache will be created and then never used...

**Comment by pearcetm** [comment]
> What do y'all think? Thank you for the in-depth analysis here!

I think this proposed implementation solves the problem nicely. It is basically what I had in mind when I said

> Logic could be added to the getCompletionCallback function to allow the callback to only be called once (instead of every time the counter drops to zero)

with the added benefit of also providing a meaningful error message when the mechanism is used inappropriately. I haven't had a chance to look at @Aiosa 's updated PR but I am in favor of this approach. Thanks for drafting it @iangilman !

**Comment by iangilman** [comment]
Fixed by #2282.
